### PR TITLE
sharding groundwork cont.

### PR DESCRIPTION
**Pull request by ymatsuda**

### CODE REVIEW

**Comment by ymatsuda** [comment]
@yingjieMiao

File: kifi-backend/modules/search/app/com/keepit/search/sharding/ShardedIndexer.scala [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  protected def resetSequenceNumberIfReindex() {
+    if (resetSequenceNumber) {
+      resetSequenceNumber = false
+      sequenceNumber = SequenceNumber.MinValue
+    }
+  }
+
+  def getIndexer(shard: Shard): T = indexShards(shard)
+
+  def update(): Int
```
Comment (yingjieMiao) [inline kifi-backend/modules/search/app/com/keepit/search/sharding/ShardedIndexer.scala:30]: todo ?

Comment (ymatsuda) [inline kifi-backend/modules/search/app/com/keepit/search/sharding/ShardedIndexer.scala:30]: an abstract method to be implemented by a subclass.

File: kifi-backend/modules/search/app/com/keepit/search/graph/URIGraphIndexer.scala [hunk 2]
```diff
@@ -48,9 +49,22 @@ class URIGraphIndexer(
...
 
+  import URIGraphIndexer.URIGraphIndexable
+
   override val commitBatchSize = 500
   private val fetchSize = commitBatchSize
 
+  override def initialSequenceNumberValue: Long = {
+    val uriGraphSeq = commitData.getOrElse(Indexer.CommitData.sequenceNumber, "-1").toLong
+    val bookmarkStoreSeq = bookmarkStore.initialSequenceNumberValue
+    if (uriGraphSeq > bookmarkStoreSeq) {
```
Comment (yingjieMiao) [inline kifi-backend/modules/search/app/com/keepit/search/graph/URIGraphIndexer.scala:60]: If both uriGraph and bookmark store are sharded, the seq num of each is taken to be the min of the sharded seq num? Comparison of the two seems unclear.

File: kifi-backend/modules/search/app/com/keepit/search/sharding/ShardedIndexer.scala [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+import com.keepit.common.db.SequenceNumber
+import com.keepit.search.index.Indexer
+
+trait ShardedIndexer[T <: Indexer[_]] {
+  val indexShards: Map[Shard, T]
+  protected val updateLock = new AnyRef
+
+  private[this] var _sequenceNumber: SequenceNumber = {
+    SequenceNumber(indexShards.valuesIterator.map(indexer => indexer.commitSequenceNumber.map(_.value).getOrElse(-1L)).min)
```
Comment (yingjieMiao) [inline kifi-backend/modules/search/app/com/keepit/search/sharding/ShardedIndexer.scala:11]: how about take max? this is related to my discussion on urigraphIndexer.

File: kifi-backend/modules/search/app/com/keepit/search/sharding/ShardedIndexer.scala (hunk 1 above)
Comment (yingjieMiao) [inline kifi-backend/modules/search/app/com/keepit/search/sharding/ShardedIndexer.scala:30]: ah.

**Comment by yingjieMiao** [comment]
looks great!

**Pull request merged**